File: ios/WhoWouldWin/Views/Components/AnimalCard.swift
import SwiftUI

struct AnimalCard: View {
    let animal: Animal
    let isSelected: Bool
    let isDisabled: Bool
    var isLocked: Bool = false
    let onTap: () -> Void

    private var accentColor: Color { Theme.categoryAccent(animal.category) }

    var body: some View {
        Button(action: {
            if !isDisabled { onTap() }
        }) {
            ZStack(alignment: .topTrailing) {
                // Card body — 3D embossed style
                VStack(spacing: 5) {
                    Group {
                        if let assetName = animal.creatureAssetName,
                           let img = UIImage(named: assetName) {
                            Image(uiImage: img)
                                .resizable()
                                .scaledToFit()
                                .frame(width: 44, height: 44)
                                .clipShape(RoundedRectangle(cornerRadius: 8))
                        } else if animal.isCustom, let url = animal.imageURL {
                            AsyncImage(url: url) { phase in
                                switch phase {
                                case .success(let img):
                                    img.resizable().scaledToFill()
                                        .frame(width: 44, height: 44)
                                        .clipShape(RoundedRectangle(cornerRadius: 8))
                                default:
                                    Text(animal.emoji).font(.system(size: 36))
                                }
                            }
                            .frame(width: 44, height: 44)
                        } else {
                            Text(animal.emoji)
                                .font(.system(size: 40))
                        }
                    }
                    .shadow(color: .black.opacity(0.3), radius: 4, x: 0, y: 3)
                    .blur(radius: isLocked ? 3 : 0)

                    Text(animal.name)
                        .font(Theme.bungee(11))
                        .foregroundColor(.white)
                        .shadow(color: .black.opacity(0.6), radius: 2, x: 0, y: 1)
                        .multilineTextAlignment(.center)
                        .lineLimit(2)
                        .minimumScaleFactor(0.65)
                        .frame(height: 28, alignment: .top)
                        .blur(radius: isLocked ? 2.5 : 0)
                }
                .frame(maxWidth: .infinity)
                .padding(.vertical, 12)
                .padding(.horizontal, 4)
                .background(
                    ZStack {
                        // Bottom 3D edge
                        RoundedRectangle(cornerRadius: 14)
                            .fill(accentColor.opacity(0.7))
                            .offset(y: 4)

                        // Main card face
                        RoundedRectangle(cornerRadius: 14)
                            .fill(Theme.categoryGradient(animal.category))

                        // Dark inset — lets category color bleed through
                        RoundedRectangle(cornerRadius: 10)
                            .fill(Color.black.opacity(0.65))
                            .padding(3)

                        // Top shine
                        VStack {
                            RoundedRectangle(cornerRadius: 12)
                                .fill(
                                    LinearGradient(
                                        colors: [Color.white.opacity(0.2), Color.white.opacity(0)],
                                        startPoint: .top, endPoint: .bottom
                                    )
                                )
                                .frame(height: 30)
                                .padding(.horizontal, 3)
                                .padding(.top, 2)
                            Spacer()
                        }
                    }
                )
                .overlay(
                    RoundedRectangle(cornerRadius: 14)
                        .stroke(
                            isSelected ? Color.white : Color.white.opacity(0.2),
                            lineWidth: isSelected ? 2.5 : 1
                        )
                )
                .shadow(
                    color: isSelected ? accentColor.opacity(0.5) : .black.opacity(0.2),
                    radius: isSelected ? 10 : 4,
                    x: 0, y: isSelected ? 4 : 3
                )

                // Lock overlay
                if isLocked {
                    ZStack {
                        RoundedRectangle(cornerRadius: 14)
                            .fill(Color.black.opacity(0.45))
                        VStack(spacing: 4) {
                            Image(systemName: "lock.fill")
                                .font(.system(size: 18, weight: .bold))
                                .foregroundColor(Theme.fantasyAccent)
                                .shadow(color: Theme.fantasyAccent.opacity(0.6), radius: 4)
                            Text("LOCKED")
                                .font(.system(size: 7, weight: .black, design: .rounded))
                                .foregroundColor(Theme.fantasyAccent.opacity(0.9))
                                .tracking(1)
                        }
                    }
                }

                // Selected X badge
                if isSelected {
                    ZStack {
                        Circle()
                            .fill(Theme.red)
                            .frame(width: 20, height: 20)
                            .shadow(color: Theme.red.opacity(0.5), radius: 3, x: 0, y: 1)
                        Image(systemName: "xmark")
                            .font(.system(size: 8, weight: .black))
                            .foregroundColor(.white)
                    }
                    .offset(x: 5, y: -5)
                }
            }
            .scaleEffect(isSelected ? 1.06 : 1.0)
            .opacity(isDisabled ? 0.32 : 1.0)
            .animation(.spring(response: 0.28, dampingFraction: 0.62), value: isSelected)
        }
        .buttonStyle(PressableButtonStyle())
        .disabled(isDisabled)
    }
}

#Preview {
    HStack(spacing: 10) {
        AnimalCard(animal: Animal(id: "lion", name: "Lion", emoji: "🦁", category: .land, pixelColor: "#D4A017", size: 4), isSelected: false, isDisabled: false, onTap: {})
        AnimalCard(animal: Animal(id: "shark", name: "Great White Shark", emoji: "🦈", category: .sea, pixelColor: "#5588AA", size: 4), isSelected: true, isDisabled: false, onTap: {})
        AnimalCard(animal: Animal(id: "eagle", name: "Bald Eagle", emoji: "🦅", category: .air, pixelColor: "#8B4513", size: 2), isSelected: false, isDisabled: true, onTap: {})
    }
    .padding()
    .background(Theme.homeBg)
}

File: android/app/src/main/java/com/whowouldin/whowouldwin/ui/components/AnimalCard.kt
package com.whowouldin.whowouldwin.ui.components

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.whowouldin.whowouldwin.model.Animal
import com.whowouldin.whowouldwin.ui.theme.BrandTheme
import com.whowouldin.whowouldwin.ui.theme.bungee

/**
 * Ports iOS `Views/Components/AnimalCard.swift`.
 *
 * 3D embossed card with category-tinted gradient, bottom "edge" offset,
 * dark inset, top shine, selection ring + X-badge, locked state with blur.
 */
@Composable
fun AnimalCard(
    animal: Animal,
    isSelected: Boolean,
    isDisabled: Boolean,
    modifier: Modifier = Modifier,
    isLocked: Boolean = false,
    onTap: () -> Unit,
) {
    val accent = BrandTheme.categoryAccent(animal.category)

    val scale by animateFloatAsState(
        targetValue = if (isSelected) 1.06f else 1f,
        animationSpec = spring(dampingRatio = 0.62f, stiffness = Spring.StiffnessMediumLow),
        label = "selScale",
    )

    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val pressScale by animateFloatAsState(
        targetValue = if (pressed && !isDisabled) 0.96f else 1f,
        animationSpec = spring(dampingRatio = Spring.DampingRatioMediumBouncy),
        label = "pressScale",
    )

    Box(
        modifier = modifier
            .scale(scale * pressScale)
            .alpha(if (isDisabled) 0.32f else 1f)
            .clickable(
                interactionSource = interaction,
                indication = null,
                enabled = !isDisabled,
                onClick = onTap,
            ),
    ) {
        // Bottom 3D edge — accent-tinted, shifted down 4dp
        Box(
            modifier = Modifier
                .matchParentSize()
                .offset(y = 4.dp)
                .background(accent.copy(alpha = 0.7f), RoundedCornerShape(14.dp)),
        )

        // Main card face
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .shadow(
                    elevation = if (isSelected) 10.dp else 4.dp,
                    shape = RoundedCornerShape(14.dp),
                    spotColor = if (isSelected) accent.copy(alpha = 0.5f) else Color.Black.copy(alpha = 0.2f),
                )
                .background(BrandTheme.categoryGradient(animal.category), RoundedCornerShape(14.dp))
                .border(
                    width = if (isSelected) 2.5.dp else 1.dp,
                    color = if (isSelected) Color.White else Color.White.copy(alpha = 0.2f),
                    shape = RoundedCornerShape(14.dp),
                )
                .padding(3.dp)
                .background(Color.Black.copy(alpha = 0.65f), RoundedCornerShape(10.dp)),
        ) {
            // Top shine
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(30.dp)
                    .padding(horizontal = 3.dp, vertical = 2.dp)
                    .background(
                        Brush.verticalGradient(
                            colors = listOf(Color.White.copy(alpha = 0.2f), Color.White.copy(alpha = 0f)),
                        ),
                        RoundedCornerShape(12.dp),
                    ),
            )

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 4.dp, vertical = 12.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(5.dp),
            ) {
                // Image / emoji — prefer bundled asset, then custom URL, then emoji
                val ctx = LocalContext.current
                val assetResId = animal.creatureAssetName?.let { name ->
                    remember(name) {
                        ctx.resources.getIdentifier(name, "drawable", ctx.packageName)
                    }
                } ?: 0

                Box(
                    modifier = Modifier
                        .size(44.dp)
                        .shadow(4.dp, RoundedCornerShape(8.dp))
                        .blur(if (isLocked) 3.dp else 0.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    when {
                        assetResId != 0 -> {
                            androidx.compose.foundation.Image(
                                painter = painterResource(id = assetResId),
                                contentDescription = animal.name,
                                contentScale = ContentScale.Fit,
                                modifier = Modifier
                                    .size(44.dp)
                                    .clip(RoundedCornerShape(8.dp)),
                            )
                        }
                        animal.isCustom && !animal.imageUrl.isNullOrBlank() -> {
                            AsyncImage(
                                model = animal.imageUrl,
                                contentDescription = animal.name,
                                contentScale = ContentScale.Crop,
                                modifier = Modifier
                                    .size(44.dp)
                                    .clip(RoundedCornerShape(8.dp)),
                            )
                        }
                        else -> {
                            Text(
                                text = animal.emoji,
                                style = bungee(40).copy(fontSize = 40.sp),
                            )
                        }
                    }
                }

                Text(
                    text = animal.name,
                    style = bungee(11),
                    color = Color.White,
                    textAlign = TextAlign.Center,
                    maxLines = 2,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(28.dp)
                        .blur(if (isLocked) 2.5.dp else 0.dp),
                )
            }
        }

        // Locked overlay
        if (isLocked) {
            Box(
                modifier = Modifier
                    .matchParentSize()
                    .background(Color.Black.copy(alpha = 0.45f), RoundedCornerShape(14.dp)),
                contentAlignment = Alignment.Center,
            ) {
                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(4.dp),
                ) {
                    Icon(
                        imageVector = Icons.Filled.Lock,
                        contentDescription = "Locked",
                        tint = BrandTheme.fantasyAccent,
                        modifier = Modifier.size(18.dp),
                    )
                    Text(
                        text = "LOCKED",
                        style = bungee(7).copy(
                            color = BrandTheme.fantasyAccent.copy(alpha = 0.9f),
                            fontWeight = FontWeight.Black,
                            letterSpacing = 1.sp,
                        ),
                    )
                }
            }
        }

        // Selected X badge — top-end
        if (isSelected) {
            Box(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .offset(x = 5.dp, y = (-5).dp)
                    .size(20.dp)
                    .shadow(3.dp, CircleShape, spotColor = BrandTheme.red.copy(alpha = 0.5f))
                    .background(BrandTheme.red, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = Icons.Filled.Close,
                    contentDescription = "Remove",
                    tint = Color.White,
                    modifier = Modifier.size(10.dp),
                )
            }
        }
    }
}
